### ISSUE TITLE
Memory consumption (even with modes enabled)

### ISSUE THREAD

**Issue by fvrmatteo** [JonathanSalwan/Triton#846]
Hi, I'm using the latest commit **ad282263f4ca59c42789c1842923cfd0e55841f8** of the branch **dev-arm32** and with the following C++ example I get around ~380 MB allocated at the end of the execution (before the process exit):

```
#include <iostream>
#include <triton/api.hpp>
#include <triton/x86Specifications.hpp>

using namespace triton;
using namespace triton::arch;

struct op {
  unsigned char*  inst;
  unsigned int    size;
};

int main() {
  /* Triton's allocations scope */
  {
    /* Init the triton context */
    triton::API api;
    api.setMode(triton::modes::ALIGNED_MEMORY, true);
    api.setMode(triton::modes::ONLY_ON_SYMBOLIZED, true);
    api.setMode(triton::modes::PC_TRACKING_SYMBOLIC, true);
    /* Set the arch */
    api.setArchitecture(ARCH_X86_64);
    /* Setup the trace */
    std::map<uint64_t, op> trace;
    trace[0x400000] = {(unsigned char *)"\x48\xC7\xC0\x20\x4E\x00\x00", 7}; // 0: mov rax, 20000
    trace[0x400007] = {(unsigned char *)"\x48\xFF\xC8", 3};                 // 7: dec rax
    trace[0x40000A] = {(unsigned char *)"\x48\x83\xF8\x00", 4};             // A: cmp rax, 0
    trace[0x40000E] = {(unsigned char *)"\x75\xF7", 2};                     // E: jne A
    /* Start from the first instruction */
    uint64_t PC = 0x400000;
    while (trace.find(PC) != trace.end()) {
      /* Build an instruction */
      triton::arch::Instruction inst;
      /* Setup opcode */
      inst.setOpcode(trace[PC].inst, trace[PC].size);
      /* Setup address */
      inst.setAddress(PC);
      /* Process everything */
      api.processing(inst);
      /* Get the next address */
      PC = api.getConcreteRegisterValue(api.getRegister(register_e::ID_REG_X86_RIP)).convert_to<uint64_t>();
    }
  }
  /* Pause and check the memory consumption */
  std::cout << "Check the memory consumption" << std::endl;
  std::cin.get();
  /* Exit */
  return 0;
}
```
The following issues https://github.com/JonathanSalwan/Triton/issues/562,  https://github.com/JonathanSalwan/Triton/issues/482 and https://github.com/JonathanSalwan/Triton/issues/319 have been closed as solved, so I'm wondering if I'm missing something or if the memory consumption issue is still present (even if in a different form).

Even assuming that the allocated memory is needed by Triton during the analysis, why is the memory still allocated after the scope reserved to the Triton code?

**Comment by fvrmatteo** [comment]
I'm linking against **Boost 1.72**, **Z3 4.8.7** and **Capstone 4.x** and using **LLVM 9.0.1** as compiler. I have different results with **Ubuntu 19.10** (showing 380 MB allocated before program exit) and **MacOS Catalina 10.15.3** (showing 19KB allocated before program exit). In both cases the memory peak during the execution is of 400MB more or less, but considering the lack of symbolisation of the registers/memory I would expect it to not allocate many nodes (as mentioned in this example https://github.com/JonathanSalwan/Triton/issues/319#issuecomment-315092176).

I can provide more context and/or tests to help tracking down the possible issue. What left me in doubt is the difference between the OSs (especially considering the same precompiled dependencies have been used).

**Comment by JonathanSalwan** [comment]
Thanks @fvrmatteo for this report and context. I will deep into this issue next week :)

**Comment by fvrmatteo** [comment]
I think I figured out the issue: if **setMode** is called before calling **setArchitecture**, the call chain **setArchitecture(..)** => **removeEngines()** resets the modes (to **nullptr**). Swapping the calls as follows fixes the problem with the memory consumption during the execution:
```
/* Init the triton context */
triton::API api;
/* Set the arch */
api.setArchitecture(ARCH_X86_64);
/* Setup the modes */
api.setMode(triton::modes::ALIGNED_MEMORY, true);
api.setMode(triton::modes::ONLY_ON_SYMBOLIZED, true);
```
I don't know if it's worth to mention this on the documentation or to come up with some kind of check to avoid others to do the same error; at the end of the day the modes are unrelated with the architecture so the ordering shouldn't be important.

Regarding the memory consumption on Linux, I think the cause is different but I would like to know if you can replicate it somehow (next week obviously :)).

**Comment by JonathanSalwan** [comment]
I quickly took a look at your PoC and I can confirm that there are two issues:

* modes are cleared when calling `setArchitecture`.
* there is a memory leak somewhere

I can reproduce the memory leak with Linux also. I suspect that the root cause is related to the `shared_ptr` destruction ordering which may be variable from an OS to another.

**Comment by JonathanSalwan** [comment]
Just to keep this thread up-to-date. We (@fvrmatteo and I) did some tests about the memory leak and we think that this is not an issue. Below another test:

```cpp
#include <iostream>
#include <triton/api.hpp>
#include <triton/x86Specifications.hpp>

using namespace triton;
using namespace triton::arch;

struct op {
  unsigned char*  inst;
  unsigned int    size;
};


void allocate(void) {
    /* Init the triton context */
    triton::API api;
    api.setMode(triton::modes::ALIGNED_MEMORY, true);
    api.setMode(triton::modes::ONLY_ON_SYMBOLIZED, true);
    api.setMode(triton::modes::PC_TRACKING_SYMBOLIC, true);
    /* Set the arch */
    api.setArchitecture(ARCH_X86_64);
    /* Setup the trace */
    std::map<uint64_t, op> trace;
    trace[0x400000] = {(unsigned char *)"\x48\xC7\xC0\x20\x4E\x00\x00", 7}; // 0: mov rax, 20000
    trace[0x400007] = {(unsigned char *)"\x48\xFF\xC8", 3};                 // 7: dec rax
    trace[0x40000A] = {(unsigned char *)"\x48\x83\xF8\x00", 4};             // A: cmp rax, 0
    trace[0x40000E] = {(unsigned char *)"\x75\xF7", 2};                     // E: jne A
    /* Start from the first instruction */
    uint64_t PC = 0x400000;
    while (trace.find(PC) != trace.end()) {
      /* Build an instruction */
      triton::arch::Instruction inst;
      /* Setup opcode */
      inst.setOpcode(trace[PC].inst, trace[PC].size);
      /* Setup address */
      inst.setAddress(PC);
      /* Process everything */
      api.processing(inst);
      /* Get the next address */
      PC = api.getConcreteRegisterValue(api.getRegister(register_e::ID_REG_X86_RIP)).convert_to<uint64_t>();
    }

  /* Pause and check the memory consumption */
  std::cout << "Check the memory consumption" << std::endl;
  std::cin.get();
}


int main() {
  allocate();
  allocate();
  allocate();
  /* Exit */
  return 0;
}
```

I allocate several times a `TritonContext` and between each allocation I check the memory consumption with `htop`. After the first allocation, I consumed 300MB of memory and this memory does not decrease at the return of the first call of `allocate()`. But when the second call appears, the memory does not increase also. It means that the process uses the memory previously allocated (probably an optimization of std allocator to avoid calls to the brk syscall).

I let @fvrmatteo describe its tests but we will probably close this issue (at least renaming it to keep the issue related to modes and `setArchitecture`).

**Comment by fvrmatteo** [comment]
Thanks for the update @JonathanSalwan, as you stated I also did additional tests (really stupid):

- I modified Triton to log all the allocated/deallocated objects (e.g. via shared_ptr) and kept track of the objects with a unique identifier and verified that after several execution of the original PoC all the objects were properly destroyed before the **std::cin.get()** pause;
- I executed the original PoC under Valgrind and no memory leaks related to Triton were detected;
- I noticed that keeping the original PoC process in a paused state and running a memory-intensive process in parallel forces the memory allocator to release the unused memory from the PoC process.

I can back @JonathanSalwan's feeling that it's the memory allocator doing some kind of optimization to avoid the allocation/deallocation of huge memory chunks used by Triton during the execution.

Although I would still keep the GitHub Issue somehow marked as "Memory consumption" because I think others may find it useful if they are analysing heavily obfuscated functions.
